### PR TITLE
OPT: repeated signals in $eq

### PR DESCRIPTION
**Pull request by Kmanfi**
Here is test module for this PR.
```
module test(input [7:0] a, input [7:0] b, output y, output x, output z, output yn, output xn, output zn, output r, output rn, output r1, output r1n, output s, output sn);
  assign y = ({a[0], a[0], 2'b00} == {2'b00, a[0], a[0]});
  assign x = ({a[0], a[0], 2'b00} == {2'b10, a[0], a[0]});
  assign z = ({2'b10, b[0], a[0]} == {a[0], a[0], 2'b00});

  assign yn = ({a[0], a[0], 2'b00} != {2'b00, a[0], a[0]});
  assign xn = ({a[0], a[0], 2'b00} != {2'b10, a[0], a[0]});
  assign zn = ({2'b10, b[0], a[0]} != {a[0], a[0], 2'b00});

  assign r = ({a[0], a[0]} == {a[0], a[0], 2'b00});
  assign rn = ({a[0], a[0]} != {a[0], a[0], 2'b00});
  assign r1 = ({a[0], a[0]} == {a[0], a[0], 2'b11});
  assign r1n = ({a[0], a[0]} != {a[0], a[0], 2'b11});

  assign s = ({a[0], a[0]} == 2'b11);
  assign sn = ({a[0], a[0]} != 2'b11);

endmodule
```

### CODE REVIEW

**Comment by Kmanfi** [comment]
Bug found. Fixing it.

**Comment by Kmanfi** [comment]
Still bug.

**Comment by Kmanfi** [comment]
Other compare ops have also room for optimisation, like this.
This could be gt = a[0];
`assign gt = ({a[0], a[0], 2'b00} > {2'b00, a[0], a[0]});`

**Comment by Kmanfi** [comment]
Here is test module for gt.
```
module gt_test(input [7:0] a, input [7:0] b, output y, output x, output z, output yn, output xn, output zn, output s, output gt);
  assign y = ({a[0], a[0], 2'b10} > {2'b00, a[0], a[0]}); // const 1 
  assign x = ({a[0], a[0], 2'b00} > {2'b10, a[0], a[0]}); // a[0]
  assign z = ({2'b10, b[0], a[0]} > {a[0], a[0], 2'b00}); // not(a[0])

  assign y = ({2'b00, a[0], a[0]} > {a[0], a[0], 2'b00}); // const 0
  assign x = ({2'b10, a[0], a[0]} > {a[0], a[0], 2'b00}); // not(a[0])
  assign z = ({a[0], a[0], 2'b00} > {2'b10, b[0], a[0]}); // not(a[0])

  assign yn = ({a[0], a[0], 2'b00} > {2'b00, a[0], a[0]}); // a[0]
  assign xn = ({a[0], a[0], 2'b00} > {2'b10, a[0], a[0]}); // a[0]
  assign zn = ({2'b10, b[0], a[0]} > {a[0], a[0], 2'b00}); // b[0] & not(a[0])

  assign s = ({a[0], a[0]} > 2'b11); // const 0

  assign gt = ({1'b1, 1'b1, a[0], 2'b11} > {1'b1, a[0], 1'b1, 1'b0, a[0]}); // const 1
endmodule
```

File: passes/opt/opt_expr.cc [hunk 1]
```diff
@@ -1251,7 +1251,145 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
...
+			SigBit polarity_bit = cell->type == "$eq" ? State::S0 : State::S1;
+
+			if (a_signed || b_signed)
+				goto next_cell;
+
+			for (int i = 0; i < a_width; i++)
+			{
+				if ((a_sig[i] ==  RTLIL::State::S0) || (a_sig[i] ==  RTLIL::State::S1))
+				{
+					if (compared_by_const.count(b_sig[i]))
```
Comment (cliffordwolf) [inline passes/opt/opt_expr.cc]: What if a_width > b_width? Then b_sig[i] can be out of range. (Same below with reversed roles for a and b.)

File: passes/opt/opt_expr.cc [hunk 2]
```diff
@@ -1251,7 +1251,145 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
...
+					for (auto ss: signals_map)
+					{
+						if ((ss == RTLIL::State::S0) || (ss == RTLIL::State::S1))
+						{
+							sig_const = ss;
+						} else
+						{
+							c_sig = ss;
+						}
+					}
```
Comment (cliffordwolf) [inline passes/opt/opt_expr.cc]: Imagine ({A,B} = {B,A}) . It has (a_width + b_width > 2) and (signals_map.size() == 2) but sig_const will be used uninitialized in the condition below.

File: passes/opt/opt_expr.cc (hunk 1 above)
Comment (Kmanfi) [reply passes/opt/opt_expr.cc]: Resized both to be equal size.

File: passes/opt/opt_expr.cc (hunk 2 above)
Comment (Kmanfi) [reply passes/opt/opt_expr.cc]: Added guard for doing this optimisation only for cells that has no x's or z's on inputs.

File: passes/opt/opt_expr.cc [hunk 3]
```diff
@@ -1251,7 +1251,165 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
...
+				{
+					if ((a_sig[i] ==  RTLIL::State::S0) || (a_sig[i] ==  RTLIL::State::S1))
+					{
+						if (compared_by_const.count(b_sig[i]))
+						{
+							if (compared_by_const.at(b_sig[i]) != a_sig[i])
+							{
+								// case where comparing signal against S0 and S1
+								log("Replacing cell `%s' in module `%s' with zero-driver.\n",
+									cell->name.c_str(), module->name.c_str());
```
Comment (cliffordwolf) [inline passes/opt/opt_expr.cc]: This always prints "zero driver" even when the output will be set to "1".

File: passes/opt/opt_expr.cc [hunk 4]
```diff
@@ -1251,7 +1251,165 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
...
+								
+								module->connect(RTLIL::SigSig(y_sig, RTLIL::SigSpec(polarity_bit, y_sig.size())));
```
Comment (cliffordwolf) [inline passes/opt/opt_expr.cc]: When polarity_bit is "1" and y_sig.size() is >1 then this will set all bits, but only the LSB should be set.

**Comment by cliffordwolf** [comment]
I think commit db6d78a does all this PR does, but with significantly fewer lines of code and without all the small subtle bugs (I hope :).

**Pull request closed without merging**